### PR TITLE
feat(config): parse pyproject.toml

### LINKED ISSUE #944: pyproject.toml (PEP 518) support
Hi! Is there a plan to support `pyproject.toml` from [PEP 518](https://www.python.org/dev/peps/pep-0518/)?

Specifically, it would be nice if `behave` could understand `[tool.behave]` section in the `pyproject.toml` file like other python libraries do (e.g. `black`, `coverage`, `isort`, `mypy`, `pytest`, ...)

### PR DESCRIPTION
**Pull request by pwoolvett**
fixes #944
See related discussion

### CODE REVIEW

**Comment by pwoolvett** [comment]
@jenisys note: current implementation does not behave (🤣 ) as your modification in  the comment from #944.

Specifically, when the cli option has`action=='append'`,  i'm requiring an array from toml instrad of  the '.splitlines'

the not so obvious implication is that for it to work, it *has* to be an array, which migh be counterintuitive:

```toml
# does not work in current implementation
[tool.behave]
default_tags = "foo"
```

```toml
# works in current implementation
[tool.behave]
default_tags = ["foo"]
```

**Comment by bittner** [comment]
Can you rebase your changes and then run `flake8` and `pylint` over the modules you have changed?

I believe, the tools will give you a few hints on formatting and such. If that reports too much you should only worry about the code parts you have changed.

**Comment by bittner** [comment]
> the not so obvious implication is that for it to work, it _has_ to be an array, which migh be counterintuitive:

@jenisys Maybe that's fine, though. In the end, the option is called in its plural form ("tags"). For me, this automatically triggers "list", also because TOML is more strict with formatting.

**Comment by jenisys** [comment]
Just a comment, related to the example with the `default_tags`.
The array/list will be replaced with a string because we will use tag-expressions v2 (that contain a logical Boolean expression as string). In addition, tag-expressions v2 allow you to specify in a more concise, easier to read and understand what you want to select.

OTHERWISE: Config Data Model if option/param is a array/list
If a CLI option is a list/array, you need to decide what is meant:
* Do you want to override the existing value(s)
* Do you want to append to the the list of existing values (from a config-file)

One way to avoid this problem is to have actually 2 options/params:
* One for the default value(s) if none are provided on CLI (override mode)
* One for the value that may also be used in the config-file (append mode)

**Comment by jenisys** [comment]
@bittner The name `--tags/tags/default_tags` is for historical reasons because you needed multiple terms for a real `tag-expression`.
The real, underlying name would be `tag-expression` (singular) that is used to select elements with these tags.
The current support for a list in the config-file was for backward compatibility.
This will be dropped in the future.

File: behave/configuration.py
Comment (pauloxnet) [inline behave/configuration.py:36]: `json` is int he standard library I think we can import it before of the `try` 
```suggestion
import json
try:
    import toml
    _TOML_AVAILABLE = True
except ImportError:
    _TOML_AVAILABLE = False
```

**Review by pauloxnet — CHANGES_REQUESTED** [review]

**Comment by pwoolvett** [comment]
> Can you rebase your changes and then run `flake8` and `pylint` over the modules you have changed?

Done, done, and done

> Just a comment, related to the example with the `default_tags`. The array/list will be replaced with a string because we will use tag-expressions v2 (that contain a logical Boolean expression as string). In addition, tag-expressions v2 allow you to specify in a more concise, easier to read and understand what you want to select.

What? how? I like the sound of that...

> 
> OTHERWISE: Config Data Model if option/param is a array/list If a CLI option is a list/array, you need to decide what is meant:
> 
>     * Do you want to override the existing value(s)
> 
>     * Do you want to append to the the list of existing values (from a config-file)
> 
> 
> One way to avoid this problem is to have actually 2 options/params:
> 
>     * One for the default value(s) if none are provided on CLI (override mode)
> 
>     * One for the value that may also be used in the config-file (append mode)

I think this last alternative sounds good, eg using `tags` for config and `extra-tags` for appending form cli. Otherwise, `tags` form cli overrides... but I'm not sure how this affects this PR yet

**Review by pauloxnet — APPROVED** [review]

**Comment by bittner** [comment]
@jenisys Can we merge this contribution? It would be a valuable change.

File: behave.ini
Comment (pauloxnet) [inline behave.ini:4]: Maybe we can preserve the alphabetical order
```suggestion
# FILE: .behaverc, behave.ini, pyproject.toml, setup.cfg, tox.ini
```

File: behave/configuration.py
Comment (pauloxnet) [inline behave/configuration.py:491]: I think first line comment one line below is more compliant.
```suggestion
    """
    Read configuration from pyproject.toml file.
```

File: behave/configuration.py
Comment (pauloxnet) [inline behave/configuration.py:562]: Maybe we can preserve the alphabetical order
```suggestion
            "behave.ini", ".behaverc", "pyproject.toml", "setup.cfg", "tox.ini"
```

File: docs/behave.rst
Comment (pauloxnet) [inline docs/behave.rst:291]: Maybe we can preserve the alphabetical order
```suggestion
Configuration files for *behave* are called either ".behaverc", "behave.ini",
"pyproject.toml", "setup.cfg" or "tox.ini" (your preference) and are located
in one of three places:
```

File: docs/behave.rst-template
Comment (pauloxnet) [inline docs/behave.rst-template:34]: Maybe we can preserve the alphabetical order
```suggestion
Configuration files for *behave* are called either ".behaverc", "behave.ini",
"pyproject.toml", "setup.cfg" or "tox.ini" (your preference) and are located
in one of three places:
```

**Review by pauloxnet — CHANGES_REQUESTED** [review]
I've reviewed again this PR and it look ready to be merged.
I only added some very minor code change suggestions.

File: behave/configuration.py
Comment (bittner) [reply behave/configuration.py:562]: Paolo, I believe adding the new configuration file to the end of the list we follow the principle of least surprise. Using alphabetical order just for the sake of aesthetics could have a dramatic impact. :thinking: 

**Example:** A developer that had added some config to `pyproject.toml` (just for fun, w/o having it working), having their productive setup in `setup.cfg` or `tox.ini`, may get a broken setup all of a sudden for no apparent reason.

Comment (pauloxnet) [reply behave/configuration.py:562]: It makes sense to me. I hope you can merge this one now. 🙂

**Comment by pauloxnet** [comment]
I think this PR is ready to be merged after quite a year of waiting. There's something to be improved that block the merge ?

File: tests/unit/test_configuration.py
Comment (bittner) [inline tests/unit/test_configuration.py:70]: This will not be portable, I fear. Please use `os.path.normpath(os.path.join(...))`. See below.

**Review by bittner — CHANGES_REQUESTED** [review]

**Comment by bittner** [comment]
@pwoolvett Can you please rebase the PR? Also, we may want to mention in the installation docs that one has to install the "toml" package for TOML support, e.g. via `pip install behave[toml]`.

@jenisys Other than that, are you fine to merge this PR? This way we would get feedback on TOML being used by everyone installing off HEAD. I guess that would be okay to do.

File: behave/configuration.py
Comment (bittner) [inline behave/configuration.py:563]: Why do we have `reversed` here? Doesn't this reverse the preference of the configuration files, i.e. make "pyproject.toml" be found & read first, and "behave.ini" only last? If yes, we probably want to place "pyproject.toml" on the top of the list to ensure we stay backward-compatible?

Also, why don't we simply denote the order as it is? e.g. 

```suggestion
        for filename in ["tox.ini", "setup.cfg", ".behaverc", "behave.ini", "pyproject.toml"]:
```

Side note: Maybe academic, but I'd prefer to have a list here instead of a tuple. Times have changed, and so have ideas about runtime optimization. Items of an identical type should be kept in a list, differing types in a tuple.

**Comment by pwoolvett** [comment]
rebased, corrected one suggestion, replied another. The ball is in your court for the moment @bittner :)

**Comment by pauloxnet** [comment]
For Python 3.11 are we using the toml module in the standard library?

**Comment by pwoolvett** [comment]
replaced `toml` with `tomllib`/`tomli`

where should that additional line be (`"tomli>=1.1.0; python_full_version < '3.11.0a7'"`)?
for the moment, it's just in `setup.py`, but i think it should be addded as well in `py.requirements/basic.txt` or `py.requirements/testing.txt`. Is there a script to do it or should I just add a line manually there?

**Comment by bittner** [comment]
There is a failing test for Python 2.7. The problem seems to be the bracket-based syntax, which is appears to be invalid for Python < 3.3 [according to StackOverflow](https://stackoverflow.com/a/16407465/202834).

@pwoolvett  Can you fix that, so we get ready for merging? :wrench: 

@jenisys I would merge this PR afterwards, if you don't object. :rocket: 
Do you want to take a look yourself? :eyes:

File: setup.py
Comment (bittner) [inline setup.py:140]: Flake8 wants us to place 2 spaces before comments.

What if we have Python >= 3.11.0a7? Neither tomli nor toml will be installed, will they? Do we have a solution for the user?

File: behave/configuration.py
Comment (bittner) [inline behave/configuration.py:508]: Why `"rb"`? Shouldn't we use read in text mode for cross-platform compatibility?

```suggestion
    with open(path) as tomlfile:
```

We shouldn't trust that configuration readers can handle Windows and Unix newlines interchangeably.

Pylint suggests to use self-explanatory variable names.

File: py.requirements/testing.txt
Comment (bittner) [inline py.requirements/testing.txt:26]: It might be possible to refer to setup.py's `[toml]` extras like this:

```suggestion
.[toml]
```

Do you want to try?

**Review by bittner — COMMENTED** [review]
Could you please also address the remaining review comments? Thank you!

**Comment by pauloxnet** [comment]
> There is a failing test for Python 2.7.

Behave is still supporting Python2 after 2 years of it's EOL date ?

I think it's time to remove support for Python2, I'm going to open a separate issue.

**Comment by pwoolvett** [comment]
> > There is a failing test for Python 2.7.
> 
> Behave is still supporting Python2 after 2 years of it's EOL date ?
> 
> I think it's time to remove support for Python2, I'm going to open a separate issue.

My thoughts exactly :rofl: 

@bittner ill take care of your suggestions later today :crossed_fingers:

**Comment by pwoolvett** [comment]
can take another shot next monday

**Comment by pauloxnet** [comment]
@pwoolvett thank you for your persistence in continuing to work on this PR.

I left a couple of absolutely non-blocking suggestions.

**Review by pauloxnet — APPROVED** [review]

**Review by bittner — APPROVED** [review]
Looks good to me.

I won't comment on two minor spacing issues. The effort in this PR is already significant. Congratulations! :rocket:

**Comment by bittner** [comment]
Tests all pass, changes look acceptable. @jenisys I'll merge this PR now.

**Comment by bittner** [comment]
@pwoolvett Thanks bunches for resisting until the end! :100: 

@pauloxnet Thank you for your help in reviewing so hard! :eyes: 

Happy holiday season, everyone! :christmas_tree:  :gift: :fireworks:

**Comment by pauloxnet** [comment]
Thank you both for this merge. Now we only need a new release. 😅

**Comment by pauloxnet** [comment]
We forgot to add a line in the changelog file I think.

**Comment by bittner** [comment]
Please open a quick PR for that, referring to this PR. We'll have this merged directly then.

**Pull request merged**
